### PR TITLE
Tel-54 create post tenant endpoint

### PR DESCRIPTION
**Pull request by AmitkumarEvry**
Insert tenant functionality
Unit test cases

### CODE REVIEW

File: src/WWTelematics.TenantService/LambdaEntrypoints.cs [hunk 1]
```diff
@@ -79,7 +79,7 @@ public ApplicationLoadBalancerResponse SingleEntityHandler(ApplicationLoadBalanc
             switch (request.HttpMethod)
             {
                 case Http.Get:
-                    return this.serviceProvider.GetService<EntityManager>().GetEntity(entityId);
+                    return this.serviceProvider.GetService<TenantManager>().GetEntity(entityId);
```
Comment (NagarajNidagundi) [inline src/WWTelematics.TenantService/LambdaEntrypoints.cs]: Rename method name

**Comment by NagarajNidagundi** [comment]
Looks good

File: src/WWTelematics.TenantService.Tests/Data/MinimalValidData.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                    {
+                        new Tenant
+                        {
+                            Id = 1,
+                            PublicId = new Guid("00000000-0000-0000-0000-000000000001"),
+                            Label = "Label 1",
+                            CreatedTime = DateTime.UtcNow,
+                            Active = true,
+                            LicenseCount = 10,
+                            ShardId = new Guid("31000000-0000-0000-0000-000000000013"),
```
Comment (davidw-foxtrax) [inline src/WWTelematics.TenantService.Tests/Data/MinimalValidData.cs]: Can we use the same shardId we're using as the default in the Manager as the default valid value here?

File: src/WWTelematics.TenantService.Tests/Application/TenantManagerTest.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+            Assert.Equal(400, albResponse.StatusCode);
+        }
+
+        /// <summary>
+        /// Test that the CreateTenant function save an ALB request if label contains | character .
+        /// Negative test case, Returns 400 response code.
+        /// </summary>
+        [Fact]
+        public void CannotSaveSingleTenantIfTextNotAlphanumeric()
```
Comment (davidw-foxtrax) [inline src/WWTelematics.TenantService.Tests/Application/TenantManagerTest.cs]: Can this test be called "CannotSaveSingleTenantIfLabelIncludesPipe"?

Other non alphanumeric characters should be allowed.

File: src/WWTelematics.TenantService/Validators/TenantValidator.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    public class TenantValidator : AbstractValidator<Tenant>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TenantValidator"/> class.
+        /// </summary>
+        public TenantValidator()
+        {
+            this.RuleFor(x => x.Label).NotEmpty().NotNull().Length(1, 31)
+                     .WithMessage("Label text length should be between 1 to 31.")
+                .Must(m => !m.Contains("|", StringComparison.CurrentCultureIgnoreCase))
```
Comment (davidw-foxtrax) [inline src/WWTelematics.TenantService/Validators/TenantValidator.cs]: This check doesn't seem to be working properly.

If label is not passed in the POST at all, a System.NullReferenceException is thrown; a 400 error is expected.

File: src/WWTelematics.TenantService/Validators/TenantValidator.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+                    .WithMessage("Label text should not conain '|'");
```
Comment (davidw-foxtrax) [inline src/WWTelematics.TenantService/Validators/TenantValidator.cs]: "contain" should have a t.

File: src/WWTelematics.TenantService/Validators/TenantValidator.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    /// <summary>
+    /// Tenant Validator.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1710:Identifiers should have correct suffix", Justification = "Validators implement IEnumerable to allow looking at rules, but are not fundamentally collections.")]
+    public class TenantValidator : AbstractValidator<Tenant>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TenantValidator"/> class.
+        /// </summary>
+        public TenantValidator()
```
Comment (davidw-foxtrax) [inline src/WWTelematics.TenantService/Validators/TenantValidator.cs]: LicenseCount is a required parameter and must be checked.

Currently, passing licenseCount = -1 works (should throw a validation error).
Not passing licenseCount works, creating a Tenant with licenseCount=0; this should throw a validation error.

File: src/WWTelematics.TenantService/Application/TenantManager.cs [hunk 7]
```diff
@@ -77,5 +78,68 @@ public ApplicationLoadBalancerResponse GetEntityList(IDictionary<string, string>
...
+            Tenant tenant;
+
+            if (string.IsNullOrEmpty(tenantRequest))
+            {
+                return ALBResponseHelper.GetDefaultALBResponse(400);
+            }
+
+            try
+            {
+                tenant = JsonConvert.DeserializeObject<Tenant>(tenantRequest);
```
Comment (davidw-foxtrax) [inline src/WWTelematics.TenantService/Application/TenantManager.cs]: The tenantRequest JSON should be encapsulated

```
{
    "tenant": {
        "label": "My Label",
        "licenseCount": 6
    }
}

File: src/WWTelematics.TenantService/Validators/TenantValidator.cs (hunk 5 above)
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService/Validators/TenantValidator.cs]: Fixed message

File: src/WWTelematics.TenantService/Validators/TenantValidator.cs (hunk 4 above)
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService/Validators/TenantValidator.cs]: Fixed validation code as suggested approach

File: src/WWTelematics.TenantService.Tests/Application/TenantManagerTest.cs (hunk 3 above)
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService.Tests/Application/TenantManagerTest.cs]: Renamed

File: src/WWTelematics.TenantService.Tests/Data/MinimalValidData.cs (hunk 2 above)
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService.Tests/Data/MinimalValidData.cs]: Changed ShardId constant value as suggested.

File: src/WWTelematics.TenantService/Application/TenantManager.cs [hunk 8]
```diff
@@ -77,5 +78,68 @@ public ApplicationLoadBalancerResponse GetEntityList(IDictionary<string, string>
...
+            catch (JsonException)
+            {
+                var error = new JsonSerializationException("Unable to parse the request body. Please submit in Json format.");
+                return ALBResponseHelper.GetDefaultALBResponse(400, error.ToString());
+            }
+
+            tenant.CreatedTime = DateTime.UtcNow;
+            tenant.PublicId = Guid.NewGuid();
+            tenant.Active = true;
+            tenant.ShardId = GetShardId();
```
[the remark this answers is not in the document]
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService/Application/TenantManager.cs]: Changed as suggested.

File: src/WWTelematics.TenantService/Validators/TenantValidator.cs (hunk 6 above)
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService/Validators/TenantValidator.cs]: Added validation for "licenseCount " property.
Also marked "licenseCount" property IsRequired in TenantContext file.

File: src/WWTelematics.TenantService/Application/TenantManager.cs (hunk 7 above)
Comment (AmitkumarEvry) [reply src/WWTelematics.TenantService/Application/TenantManager.cs]: Introduced a TenantWrapper class And using Tenant class as property inside TenantWrapper class to achieve provided json format as input/output.

**Pull request merged**